src/p2p/swarm.rs

```diff
@@ -112,10 +119,12 @@ impl SwarmApi {
             ConnectionTarget::Addr(ref addr) => NetworkBehaviourAction::DialAddress {
                 address: addr.clone(),
             },
-            ConnectionTarget::PeerId(ref id) => NetworkBehaviourAction::DialPeer {
-                peer_id: id.clone(),
-                condition: DialPeerCondition::Disconnected,
-            },
+            ConnectionTarget::PeerId(ref id) | ConnectionTarget::PeerWithAddrs(ref id, _) => {
+                NetworkBehaviourAction::DialPeer {
+                    peer_id: id.clone(),
+                    condition: DialPeerCondition::Disconnected,
+                }
+            }
```

Comment:
Just thought about this: what if we are already connected to the PeerId? I think this impl will hang indefinitely until there's a disconnect. I think the way to go would be an "ready subscription", perhaps?

Reply:
Perhaps this could be outside of this PR.